pkg/lexer/lexer.go

```diff
@@ -56,64 +63,133 @@ func readInt(l *lex.State) string {
 
 func readNumber(l *lex.State) lex.StateFn {
 	l.StartToken(l.Pos())
+	// By default we are here if either first char is
+	// [0-9] -- which means it is positive
+	// +     -- which means it is positive
+	// -     -- which means it is negative
+	//
+	// So there is a single case when it is positive
+	mul := int64(1)
+	switch r := l.Peek(); r {
+	case '+':
+		l.Next()
+	case '-':
+		l.Next()
+		mul = -1
+	}
+
 	value := readInt(l)
 
 	switch r := l.Peek(); {
 	case r == '.':
 		l.Next()
 		value += string(r) + readInt(l)
 
-		if r = l.Peek(); !(r == lex.EOF || unicode.IsSpace(r) || r == ')') {
-			l.Errorf(l.TokenPos(), "Real numbers should end with space, eof, or right bracket, not with '%v'", string(r))
+		if r = l.Peek(); !isAtomEnd(r) {
+			l.Errorf(l.TokenPos(), "Real numbers shouldn't end with `%v'", string(r))
 			return nil
 		}
 
 		f, err := strconv.ParseFloat(value, 64)
+		f *= float64(mul)
 		if err != nil {
 			l.Errorf(l.TokenPos(), "Error parsing real number: %v", err)
 		} else {
-			l.Emit(l.TokenPos(), tok.REAL, &f)
+			l.Emit(l.TokenPos(), tok.REAL, f)
 		}
-	case unicode.IsSpace(r) || r == lex.EOF || r == ')':
+	case isAtomEnd(r):
 		i, err := strconv.ParseInt(value, 10, 64)
+		i *= mul
+		new_i := int(i)
 		if err != nil {
 			l.Errorf(l.TokenPos(), "Error parsing integer: %v", err)
 		} else {
-			l.Emit(l.TokenPos(), tok.INT, &i)
+			l.Emit(l.TokenPos(), tok.INT, new_i)
 		}
 	default:
-		l.Errorf(l.TokenPos(), "Integers should end with space, eof, or right bracket, not with `%v'", string(r))
+		l.Errorf(l.TokenPos(), "Integers shouldn't end with `%v'", string(r))
+	}
+
+	return nil
+}
+
+func readRawChar(l *lex.State) *rune {
+	r := l.Next()
+	if r == '\\' {
+		switch r := l.Next(); r {
+		case 'r':
+			r = '\r'
+		case 'n':
+			r = '\n'
+		case '\\':
+			r = '\\'
+		case lex.EOF:
+			l.Errorf(l.Pos(), "Expected character after backslash, but got EOF")
+			return nil
+		default:
+			l.Errorf(l.Pos(), "Unknown escape char `%v'", string(r))
+			return nil
+		}
+	}
+	return &r
+}
+
+func readChar(l *lex.State) lex.StateFn {
+	l.StartToken(l.Pos())
+	assertChar(l, '\'')
+	r := readRawChar(l)
+	if r == nil {
+		return nil
+	}
+	if got, assert := assertChar(l, '\''); assert {
+		l.Errorf(l.TokenPos(), "Character symbol should end with single quote, not with `%v'", string(got))
 	}
+	l.Emit(l.TokenPos(), tok.LETTER, *r)
+	return nil
+}
 
+func readString(l *lex.State) lex.StateFn {
```

Comment:
```suggestion
func readString(l *lex.State) lex.StateFn {
	l.StartToken(l.Pos())
```
Apparently it's called everywhere, maybe need here as well?

Reply:
Same as for above